rewrite-xml/src/main/java/org/openrewrite/xml/ChangeNamespaceValue.java

```diff
@@ -93,47 +92,57 @@ public ChangeNamespaceValue(@Nullable @JsonProperty("elementName") String elemen
     public TreeVisitor<?, ExecutionContext> getVisitor() {
         return new XmlIsoVisitor<ExecutionContext>() {
 
-            String version = null;
-
             @Override
             public Xml.Tag visitTag(Xml.Tag tag, ExecutionContext ctx) {
                 Xml.Tag t = super.visitTag(tag, ctx);
-                this.version = null;
-                if (elementName == null || new XPathMatcher(elementName).matches(getCursor())) {
-                    // find versions
-                    t = t.withAttributes(ListUtils.map(t.getAttributes(), this::visitChosenElementAttribute));
-                }
 
-                if (this.version != null) {
-                    //change namespace
-                    t = t.withAttributes(ListUtils.map(t.getAttributes(), this::visitChosenElementAttribute));
+                if (matchesElementName(getCursor()) && matchesVersion(t)) {
+                    t = t.withAttributes(ListUtils.map(t.getAttributes(), this::maybeReplaceNamespaceAttribute));
                 }
+
                 return t;
             }
 
-            public Xml.Attribute visitChosenElementAttribute(Xml.Attribute attribute) {
-                boolean isXmlnsAttr = (searchAllNamespaces && attribute.getKeyAsString().startsWith(XMLNS_PREFIX) ||
-                        !searchAllNamespaces && attribute.getKeyAsString().equals(XMLNS_PREFIX));
-                boolean isVersionAttr = attribute.getKeyAsString().startsWith(VERSION_PREFIX);
-                if (oldValue != null && (!isXmlnsAttr || !attribute.getValueAsString().equals(oldValue))) {
-                    return attribute;
-                }
+            private boolean matchesElementName(Cursor cursor) {
+                return elementName == null || new XPathMatcher(elementName).matches(cursor);
```

Comment:
Can we extract this `XPathMatcher` to a fields and reuse it? Now it's created for every tag we visit.